sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala

```diff
@@ -118,17 +121,21 @@ object CollationTypeCasts extends TypeCoercionRule {
           )
       // Only implicit or default collations present
       case 0 =>
-        val implicitTypes = expr.map(_.dataType)
+        val implicitTypes = expr.filter {
+            case Literal(_, _: StringType) => false
+            case Cast(child, _: StringType, _, _) => child.dataType.isInstanceOf[StringType]
```

Comment:
Let's also check `cast.getTagValue(Cast.USER_SPECIFIED_CAST).isEmpty` to make sure it's not a user-specified cast.

Reply:
oh wait, if users use scala/python DataFrame to cast a column to a StringType with specific collation, should it be explicit collation?

Reply:
Same for SQL: `CAST(col AS STRING WITH ...)`